### PR TITLE
Changing the handle we use for IPAM

### PR DESCRIPTION
**Pull request by heschlie**
## Description

Instead of just using the workload, we use the container network
namespace and the container ID combined to generate a unique handle to
track allocations.

Some questions before the WIP is removed:
- Should we be trimming off the path portion of the container network ns? It seems like so but I haven't done that yet.
- There was mention of separate code paths for k8s vs others, and it seems like they both lead back to `calico-ipam.go` so I feel like I've probably overlooked something in the code path for CNI
- If we trim out the path portion of the container net ns (and maybe even if not) I don't think there is any escaping that needs to be done
- Should we bother to check in `utils.GetHandleId()` for the container ID and network namespace as they are supposed to be a required part of the CNI spec (provided the orchestrators are all following it properly)?

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: calico_cni_k8s_test.go [hunk 1]
```diff
@@ -969,7 +969,7 @@ var _ = Describe("CalicoCni", func() {
 
 				checkIPAMReservation := func() {
 					// IPAM reservation should still be in place.
-					ipamIPs, err := calicoClient.IPAM().IPsByHandle(workloadName)
+					ipamIPs, err := calicoClient.IPAM().IPsByHandle(contNs.Path() + "." + containerID)
```
Comment (caseydavenport) [inline calico_cni_k8s_test.go]: `contNs.path()` isn't what we want to use here.

The key information should be some form of `(cni_network_name, containerID)`

File: utils/utils.go [hunk 2]
```diff
@@ -171,6 +171,10 @@ func GetIdentifiers(args *skel.CmdArgs) (workloadID string, orchestratorID strin
 	return workloadID, orchestratorID, nil
 }
 
+func GetHandleId(args *skel.CmdArgs) (handleId string, err error) {
+	return fmt.Sprintf("%s.%s", args.Netns, args.ContainerID), nil
```
Comment (caseydavenport) [inline utils/utils.go]: again, the netns isn't the thing we want to use - it's the network name.

File: ipam/calico-ipam.go [hunk 3]
```diff
@@ -97,11 +97,11 @@ func cmdAdd(args *skel.CmdArgs) error {
 		return err
 	}
 
-	workloadID, _, err := utils.GetIdentifiers(args)
+	handleId, err := utils.GetHandleId(args)
 	if err != nil {
 		return err
 	}
-	logger := utils.CreateContextLogger(workloadID)
+	logger := utils.CreateContextLogger(handleId)
```
Comment (fasaxc) [inline ipam/calico-ipam.go]: Please can we have a logger that has the workload ID *and* the handle ID in its context?  The container ID is just a random string so the really useful thing is the combination of the two.

File: ipam/calico-ipam.go [hunk 4]
```diff
@@ -222,18 +222,34 @@ func cmdDel(args *skel.CmdArgs) error {
...
+		return err
+	}
 
+	// We also need to handle the old way by workloadID, to not leak IPs in existing installs.
 	logger.Info("Releasing address using workloadID")
-	if err := calicoClient.IPAM().ReleaseByHandle(workloadID); err != nil {
+	if err := calicoClient.IPAM().ReleaseByHandle(workloadId); err != nil {
 		if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
-			logger.WithField("workloadId", workloadID).Warn("Asked to release address but it doesn't exist. Ignoring")
+			logger.WithField("workloadId", workloadId).Warn("Asked to release address but it doesn't exist. Ignoring")
```
Comment (fasaxc) [inline ipam/calico-ipam.go]: If we succeeded in deleting the key above, then we should suppress this warning.  Going forward, we only expect the above handle ID delete to succeed.

File: ipam/calico-ipam.go (hunk 3 above)
Comment (fasaxc) [reply ipam/calico-ipam.go]: I'm also not a fan of putting _everything_ in a utils package, what's wrong with `logger := log.WithField("handle", handleID)`  It's just a short and it saves a trip to the other file to see what's going on.

File: utils/utils.go [hunk 5]
```diff
@@ -171,6 +171,10 @@ func GetIdentifiers(args *skel.CmdArgs) (workloadID string, orchestratorID strin
 	return workloadID, orchestratorID, nil
 }
 
+func GetHandleId(args *skel.CmdArgs) (handleId string, err error) {
```
Comment (fasaxc) [inline utils/utils.go]: Nit: go naming convention is `GetHandleID`

File: ipam/calico-ipam.go [hunk 6]
```diff
@@ -222,18 +222,34 @@ func cmdDel(args *skel.CmdArgs) error {
...
 	}
 
-	logger := utils.CreateContextLogger(workloadID)
+	// Release the IP address by using the workloadID, to not leak IPs in existing installs.
+	workloadId, _, err := utils.GetIdentifiers(args)
+	if err != nil {
+		return err
+	}
+
+	logger := utils.CreateContextLogger(handleId)
```
Comment (fasaxc) [inline ipam/calico-ipam.go]: Let's have a logger with both workloadID and handle here.  It's super useful to be able to grep the kubelet log for the workload ID and see all the handles or vice versa.

**Comment by heschlie** [comment]
In the tests I hardcoded the network name, as we're not returning it anywhere at the moment. I did notice that in `RunCNIPluginWithId` in the test_utils that we have a TODO to use `NetConf` instead of calling the CNI plugin, should we go ahead and switch now since we could use some info from it, and add the `NetConf` struct to the returned items (maybe trimming out some of them that are included in the struct as well)?

File: ipam/calico-ipam.go [hunk 7]
```diff
@@ -222,24 +219,43 @@ func cmdDel(args *skel.CmdArgs) error {
 		return err
 	}
 
-	// Release the IP address by using the handle - which is workloadID.
+	// Release the IP address by using the handleID.
+	handleID := fmt.Sprintf("%s.%s", conf.Name, args.ContainerID)
```
Comment (caseydavenport) [inline ipam/calico-ipam.go]: Are we worried about `conf.Name` containing invalid characters?

The CNI spec says this field is a "string" but doesn't give any more detail than that. Do we actually need to hash this or do some escaping to ensure we don't try to write a bad value?

Comment (heschlie) [reply ipam/calico-ipam.go]: I think we do, if we don't really care how pretty it is hashing it seems like the better option to me, but if we want it to look sensible we should escape it, do we already have a function for this in libcalico somewhere?

Comment (heschlie) [reply ipam/calico-ipam.go]: I went with b64 encoding the network name, once the tests pass locally I'll push up the code for the next review.

**Comment by heschlie** [comment]
@caseydavenport @fasaxc Another push for your review.

File: ipam/calico-ipam.go [hunk 8]
```diff
@@ -222,24 +222,46 @@ func cmdDel(args *skel.CmdArgs) error {
 		return err
 	}
 
-	// Release the IP address by using the handle - which is workloadID.
+	// Release the IP address by using the handleID.
```
Comment (caseydavenport) [inline ipam/calico-ipam.go]: @heschlie I find the ordering of the current logic to be pretty confusing, and a bit hard to follow. Roughly, we've got:

- calculate the handle ID
- calculate the workloadID
- attempt release using the handle ID
  - unset a bool if IP not assigned
- attempt release using the workload ID, check boolean, emit error based on value.
  - only log if does not exist, and the previous bool was unset. 

I'm not a huge fan of using a boolean here for the sole purpose of hiding a log.

What about a flow like this?

- calculate handle ID
- attempt release using handle ID
  - if does not exist error:
    - calculate workloadID
    - attempt release using the workload ID
      - if does not exist error, log warning and return nil
  - if another type of error:
    - log error and return it

We can encapsulate the legacy workloadID logic into a helper function to make this code a bit cleaner.

Comment (heschlie) [reply ipam/calico-ipam.go]: Originally I was thinking that there might be a situation where both the `handleID` and `workloadId` would exist at the same time, so we would need to call both (though admittedly could have been written better) but I'm less certain of that now, so I think encapsulating the `workloadId` call  is the way to go.

Comment (caseydavenport) [reply ipam/calico-ipam.go]: As discussed offline, let's still release both, but I think we can clean up the logic a bit.

File: ipam/calico-ipam.go [hunk 9]
```diff
@@ -222,24 +222,41 @@ func cmdDel(args *skel.CmdArgs) error {
...
-	logger.Info("Released address using workloadID")
-	return nil
+	// We also need to handle the old way by workloadID, to not leak IPs in existing installs.
+	workloadLogger.Info("Releasing address using workloadID")
+	if err := calicoClient.IPAM().ReleaseByHandle(workloadID); err != nil {
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			return err
+		}
+	} else {
+		workloadLogger.Info("Released address using workloadId")
```
Comment (caseydavenport) [inline ipam/calico-ipam.go]: nit: `workloadID`

File: ipam/calico-ipam.go [hunk 10]
```diff
@@ -222,24 +222,41 @@ func cmdDel(args *skel.CmdArgs) error {
...
 	}
+	workloadLogger := utils.CreateContextLogger(workloadID)
 
-	logger.Info("Released address using workloadID")
-	return nil
+	// We also need to handle the old way by workloadID, to not leak IPs in existing installs.
+	workloadLogger.Info("Releasing address using workloadID")
+	if err := calicoClient.IPAM().ReleaseByHandle(workloadID); err != nil {
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			return err
```
Comment (caseydavenport) [inline ipam/calico-ipam.go]: Might be worth an error log in this branch.

File: ipam/calico-ipam.go [hunk 11]
```diff
@@ -222,24 +222,41 @@ func cmdDel(args *skel.CmdArgs) error {
...
+		}
```
Comment (caseydavenport) [inline ipam/calico-ipam.go]: And a debug log for the case where we get a `ResourceDoesNotExist` error.

Info is too high because this will be the mainline case, but debug should be fine and useful for debugging any potential issues with this code.

File: utils/utils.go [hunk 12]
```diff
@@ -171,6 +175,16 @@ func GetIdentifiers(args *skel.CmdArgs) (workloadID string, orchestratorID strin
 	return workloadID, orchestratorID, nil
 }
 
+func GetHandleID(netName string, containerID string) (string, error) {
```
Comment (caseydavenport) [inline utils/utils.go]: It would be really useful to include a log in this function which correlates the netname, containerID, and the generated handle ID.

Do we get that anywhere else? Need to make sure that shows up both in ADD and DEL code paths.

Comment (heschlie) [reply utils/utils.go]: Added a context logger here, not sure if it is exactly what you had in mind but it should get the idea across

**Comment by heschlie** [comment]
@caseydavenport @fasaxc ready for another look.

File: ipam/calico-ipam.go [hunk 13]
```diff
@@ -222,24 +222,45 @@ func cmdDel(args *skel.CmdArgs) error {
...
+	logger := utils.CreateContextLogger(handleID)
 
-	logger.Info("Releasing address using workloadID")
-	if err := calicoClient.IPAM().ReleaseByHandle(workloadID); err != nil {
+	logger.Info("Releasing address using handleID")
+	if err := calicoClient.IPAM().ReleaseByHandle(handleID); err != nil {
 		if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
-			logger.WithField("workloadId", workloadID).Warn("Asked to release address but it doesn't exist. Ignoring")
-			return nil
+			logger.Warn("Asked to release address but it doesn't exist. Ignoring")
```
Comment (caseydavenport) [inline ipam/calico-ipam.go]: nit: you could/should reduce this down to a single branch by using `!ok` in the check above - then we don't need the "else" branch, because we'd just return in the error case.

File: ipam/calico-ipam.go [hunk 14]
```diff
@@ -222,24 +222,45 @@ func cmdDel(args *skel.CmdArgs) error {
...
 		return err
 	}
+	workloadLogger := utils.CreateContextLogger(workloadID)
 
-	logger.Info("Released address using workloadID")
-	return nil
+	// We also need to handle the old way by workloadID, to not leak IPs in existing installs.
+	workloadLogger.Info("Releasing address using workloadID")
+	if err := calicoClient.IPAM().ReleaseByHandle(workloadID); err != nil {
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
```
Comment (caseydavenport) [inline ipam/calico-ipam.go]: Same here - if you put the error case first, then you can get rid of the "else" branch.

File: utils/utils.go [hunk 15]
```diff
@@ -171,6 +175,23 @@ func GetIdentifiers(args *skel.CmdArgs) (workloadID string, orchestratorID strin
...
+		"ContainerID": containerID,
+	})
+	hasher := md5.New()
+	_, err := hasher.Write([]byte(netName))
+	if err != nil {
+		return "", err
+	}
+	hashedName := hex.EncodeToString(hasher.Sum(nil))
+	handleID := fmt.Sprintf("%s.%s", hashedName, containerID)
+	logger.Info("Generated handleID(%s) from hashed network name and containerID", handleID)
```
Comment (caseydavenport) [inline utils/utils.go]: Just a nit here - let's just use `WithFields` for all of them to make it consistent. 

Then, the log text should just be "Generated IPAM handle"

Also, it doesn't include the workload - we should pass in the main logger in order to get that context.

And finally, I think it should be a debug log.

File: ipam/calico-ipam.go [hunk 16]
```diff
@@ -222,24 +227,42 @@ func cmdDel(args *skel.CmdArgs) error {
...
+
+	// We also need to handle the old way by workloadID, to not leak IPs in existing installs.
+	workloadLogger := utils.CreateContextLogger(workloadID)
+	workloadLogger.Info("Releasing address using workloadID")
 	if err := calicoClient.IPAM().ReleaseByHandle(workloadID); err != nil {
-		if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
-			logger.WithField("workloadId", workloadID).Warn("Asked to release address but it doesn't exist. Ignoring")
-			return nil
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			workloadLogger.Errorf("Failed to release address: %v", err)
```
Comment (fasaxc) [inline ipam/calico-ipam.go]: Please use `workloadLogger.WithError(err).Error("Failed to release address.")`

File: utils/utils.go [hunk 17]
```diff
@@ -171,6 +175,23 @@ func GetIdentifiers(args *skel.CmdArgs) (workloadID string, orchestratorID strin
 	return workloadID, orchestratorID, nil
 }
 
+func GetHandleID(netName string, containerID string, workload string) (string, error) {
+	hasher := md5.New()
```
Comment (fasaxc) [inline utils/utils.go]: Don't use MD5, it's broken.  If you need a crypto hash, reach for a SHA-2 family function.

Comment (fasaxc) [reply utils/utils.go]: Do we actually need to hash the network name?  It'd be really nice to avoid that if we can since it makes it a real pain to correlate later.

Comment (heschlie) [reply utils/utils.go]: @fasaxc this wasn't for security, it was just to ensure we ended up with only valid characters in the network name portion, as the CNI spec only dictates that it is a string. My thought was instead of trying to escape all invalid characters we just hash it as this value isn't presented to users (correct me if I'm wrong there).

If we really feel it would be helpful to store this value in a human readable format we can do that, we just need to escape the string.

Comment (heschlie) [reply utils/utils.go]: @fasaxc @caseydavenport thinking about this a bit more, are we worried about invalid values in `conf.Name`? Is the `HandleID` used anywhere but the IPAM code?

File: ipam/calico-ipam.go [hunk 18]
```diff
@@ -222,24 +227,42 @@ func cmdDel(args *skel.CmdArgs) error {
...
+
+	// We also need to handle the old way by workloadID, to not leak IPs in existing installs.
+	workloadLogger := utils.CreateContextLogger(workloadID)
+	workloadLogger.Info("Releasing address using workloadID")
 	if err := calicoClient.IPAM().ReleaseByHandle(workloadID); err != nil {
-		if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
-			logger.WithField("workloadId", workloadID).Warn("Asked to release address but it doesn't exist. Ignoring")
-			return nil
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			logger.WithError(err).Error("Failed to release address")
```
Comment (caseydavenport) [inline ipam/calico-ipam.go]: Shouldn't this be `workloadLogger`?

File: utils/utils.go [hunk 19]
```diff
@@ -171,6 +171,17 @@ func GetIdentifiers(args *skel.CmdArgs) (workloadID string, orchestratorID strin
 	return workloadID, orchestratorID, nil
 }
 
+func GetHandleID(netName string, containerID string, workload string) (string, error) {
+	handleID := fmt.Sprintf("%s.%s", netName, containerID)
+	log.WithFields(log.Fields{
+		"Network":     netName,
+		"ContainerID": containerID,
+		"Workload":    workload,
```
Comment (caseydavenport) [inline utils/utils.go]: Log is missing the handle itself :)

File: ipam/calico-ipam.go (hunk 18 above)
Comment (heschlie) [reply ipam/calico-ipam.go]: argh, copy paste problems

File: utils/utils.go (hunk 19 above)
Comment (heschlie) [reply utils/utils.go]: Should that be part of the with fields or in the log message itself?

Comment (caseydavenport) [reply utils/utils.go]: fields
